File: apps/api/src/payments/daraja.client.ts
import { Injectable, Logger, ServiceUnavailableException } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";

export interface StkPushRequest {
  phone: string; // MSISDN, e.g. 2547XXXXXXXX
  amount: number;
  accountReference: string; // e.g. the booking's public reference
  transactionDesc: string;
  callbackUrl: string;
}

export interface StkPushResponse {
  merchantRequestId: string;
  checkoutRequestId: string;
  responseCode: string;
  responseDescription: string;
}

/**
 * Thin wrapper around Safaricom's Daraja API. Every credential this needs
 * (`DARAJA_CONSUMER_KEY`, `DARAJA_CONSUMER_SECRET`, `DARAJA_PASSKEY`,
 * `DARAJA_SHORTCODE`) is `Information Required` until a real Daraja
 * account is provisioned — see apps/api/.env.example. Rather than fake a
 * successful response when unconfigured, every method refuses clearly,
 * the same pattern OtpService uses for its own missing SMS credential.
 */
@Injectable()
export class DarajaClient {
  private readonly logger = new Logger(DarajaClient.name);

  constructor(private readonly configService: ConfigService) {}

  private isConfigured(): boolean {
    const key = this.configService.get<string>("DARAJA_CONSUMER_KEY");
    return Boolean(key && key !== "Information Required");
  }

  private baseUrl(): string {
    const env = this.configService.get<string>("DARAJA_ENV") ?? "sandbox";
    return env === "production"
      ? "https://api.safaricom.co.ke"
      : "https://sandbox.safaricom.co.ke";
  }

  private async getAccessToken(): Promise<string> {
    const consumerKey = this.configService.get<string>("DARAJA_CONSUMER_KEY");
    const consumerSecret = this.configService.get<string>("DARAJA_CONSUMER_SECRET");
    const credentials = Buffer.from(`${consumerKey}:${consumerSecret}`).toString("base64");

    const response = await fetch(`${this.baseUrl()}/oauth/v1/generate?grant_type=client_credentials`, {
      headers: { Authorization: `Basic ${credentials}` },
    });

    if (!response.ok) {
      throw new ServiceUnavailableException("Could not authenticate with the M-Pesa Daraja API.");
    }
    const data = (await response.json()) as { access_token: string };
    return data.access_token;
  }

  async initiateStkPush(request: StkPushRequest): Promise<StkPushResponse> {
    if (!this.isConfigured()) {
      throw new ServiceUnavailableException(
        "M-Pesa payment collection is not yet configured. Set DARAJA_CONSUMER_KEY, " +
          "DARAJA_CONSUMER_SECRET, DARAJA_PASSKEY, and DARAJA_SHORTCODE (Part J) before enabling payments."
      );
    }

    const shortcode = this.configService.get<string>("DARAJA_SHORTCODE");
    const passkey = this.configService.get<string>("DARAJA_PASSKEY");
    const timestamp = new Date().toISOString().replace(/[^0-9]/g, "").slice(0, 14);
    const password = Buffer.from(`${shortcode}${passkey}${timestamp}`).toString("base64");

    const accessToken = await this.getAccessToken();

    const response = await fetch(`${this.baseUrl()}/mpesa/stkpush/v1/processrequest`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        BusinessShortCode: shortcode,
        Password: password,
        Timestamp: timestamp,
        TransactionType: "CustomerPayBillOnline",
        Amount: Math.round(request.amount),
        PartyA: request.phone,
        PartyB: shortcode,
        PhoneNumber: request.phone,
        CallBackURL: request.callbackUrl,
        AccountReference: request.accountReference,
        TransactionDesc: request.transactionDesc,
      }),
    });

    if (!response.ok) {
      const body = await response.text();
      this.logger.error(`Daraja STK push failed: ${body}`);
      throw new ServiceUnavailableException("M-Pesa could not process this payment request.");
    }

    const data = (await response.json()) as any;
    return {
      merchantRequestId: data.MerchantRequestID,
      checkoutRequestId: data.CheckoutRequestID,
      responseCode: data.ResponseCode,
      responseDescription: data.ResponseDescription,
    };
  }
}

File: apps/api-py/app/services/daraja_client.py
"""
Direct port of payments/daraja.client.ts.

Every credential this needs (DARAJA_CONSUMER_KEY, DARAJA_CONSUMER_SECRET,
DARAJA_PASSKEY, DARAJA_SHORTCODE) is unset until a real Daraja account is
provisioned. Rather than fake a successful response when unconfigured,
every method refuses clearly, the same pattern the other notification
clients use for their own missing credentials.
"""

import base64
import logging
from datetime import datetime, timezone
from typing import TypedDict

import httpx
from fastapi import HTTPException

from app.core.config import settings

logger = logging.getLogger(__name__)


class StkPushRequest(TypedDict):
    phone: str
    amount: float
    account_reference: str
    transaction_desc: str
    callback_url: str


class StkPushResponse(TypedDict):
    merchant_request_id: str
    checkout_request_id: str
    response_code: str
    response_description: str


def _is_configured() -> bool:
    key = settings.daraja_consumer_key
    return bool(key and key != "Information Required")


def _base_url() -> str:
    return "https://api.safaricom.co.ke" if settings.daraja_env == "production" else "https://sandbox.safaricom.co.ke"


async def _get_access_token() -> str:
    credentials = base64.b64encode(
        f"{settings.daraja_consumer_key}:{settings.daraja_consumer_secret}".encode()
    ).decode()

    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{_base_url()}/oauth/v1/generate?grant_type=client_credentials",
            headers={"Authorization": f"Basic {credentials}"},
        )
    if response.status_code >= 400:
        raise HTTPException(status_code=503, detail="Could not authenticate with the M-Pesa Daraja API.")
    return response.json()["access_token"]


async def initiate_stk_push(request: StkPushRequest) -> StkPushResponse:
    if not _is_configured():
        raise HTTPException(
            status_code=503,
            detail=(
                "M-Pesa payment collection is not yet configured. Set DARAJA_CONSUMER_KEY, "
                "DARAJA_CONSUMER_SECRET, DARAJA_PASSKEY, and DARAJA_SHORTCODE before enabling payments."
            ),
        )

    shortcode = settings.daraja_shortcode
    passkey = settings.daraja_passkey
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%d%H%M%S")
    password = base64.b64encode(f"{shortcode}{passkey}{timestamp}".encode()).decode()

    access_token = await _get_access_token()

    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{_base_url()}/mpesa/stkpush/v1/processrequest",
            headers={"Authorization": f"Bearer {access_token}", "Content-Type": "application/json"},
            json={
                "BusinessShortCode": shortcode,
                "Password": password,
                "Timestamp": timestamp,
                "TransactionType": "CustomerPayBillOnline",
                "Amount": round(request["amount"]),
                "PartyA": request["phone"],
                "PartyB": shortcode,
                "PhoneNumber": request["phone"],
                "CallBackURL": request["callback_url"],
                "AccountReference": request["account_reference"],
                "TransactionDesc": request["transaction_desc"],
            },
        )

    if response.status_code >= 400:
        logger.error("Daraja STK push failed: %s", response.text)
        raise HTTPException(status_code=503, detail="M-Pesa could not process this payment request.")

    data = response.json()
    return {
        "merchant_request_id": data["MerchantRequestID"],
        "checkout_request_id": data["CheckoutRequestID"],
        "response_code": data["ResponseCode"],
        "response_description": data["ResponseDescription"],
    }
